### PR TITLE
Feature/appointments simple search

### PR DESCRIPTION
**Pull request by dimitrijevic12**

### CODE REVIEW

File: PatientWebApplication/HealthClinic.DAL/Adapters/AppointmentAdapter.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+        }
+
+        public List<AppointmentDto> ConvertAppointmentListToAppointmentDtoList(List<DoctorAppointment> appointments)
+        {
+            List<AppointmentDto> appointmentsDto = new List<AppointmentDto>();
+            AppointmentAdapter adapter = new AppointmentAdapter();
+            foreach (DoctorAppointment appointment in appointments)
+            {
+                appointmentsDto.Add(adapter.AppointmentToAppointmenDto(appointment));
+            }
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Adapters/AppointmentAdapter.cs:23]: Refact in a special method with a meaningful name.

File: PatientWebApplication/HealthClinic.DAL/Adapters/OperationAdapter.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+        }
+
+        public List<OperationDto> ConvertOperationListToOperationDtoList(List<Operation> operations)
+        {
+            List<OperationDto> operationsDto = new List<OperationDto>();
+            OperationAdapter adapter = new OperationAdapter();
+            foreach (Operation operation in operations)
+            {
+                operationsDto.Add(adapter.OperationToOperationDto(operation));
+            }
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Adapters/OperationAdapter.cs:23]: Refact in a special method with a meaningful name.

File: PatientWebApplication/HealthClinic.DAL/Adapters/OperationAdapter.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Adapters/OperationAdapter.cs:24]: Delete the unnecessarily added blank line.

File: PatientWebApplication/HealthClinic.DAL/Adapters/AppointmentAdapter.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Adapters/AppointmentAdapter.cs:24]: Delete the unnecessarily added blank line.

File: PatientWebApplication/HealthClinic.DAL/Contoller/AppointmentController.cs [hunk 5]
```diff
@@ -39,9 +40,10 @@ public List<DoctorAppointment> GetAll()
             return regularAppointmentService.GetAll();
         }
 
-        public DoctorAppointment GetByid(int id)
+        //        public DoctorAppointment GetByid(int id)
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Contoller/AppointmentController.cs:43]: Delete the commented line of code.

File: PatientWebApplication/HealthClinic.DAL/Contoller/AppointmentController.cs [hunk 6]
```diff
@@ -39,9 +40,10 @@ public List<DoctorAppointment> GetAll()
...
+        public void GetByid(int id)
         {
-            return regularAppointmentService.GetByid(id);
+ //           return regularAppointmentService.GetByid(id);
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Contoller/AppointmentController.cs:46]: Delete the commented line of code.

File: PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs [hunk 7]
```diff
@@ -55,7 +58,68 @@ public Boolean isTermNotAvailable(DoctorUser doctor, TimeSpan start, TimeSpan en
...
+
+        public List<Operation> SimpleSearchOperations(AppointmentReportSearchDto appointmentReportSearchDto)
+        {
+            List<Operation> operations = GetOperationsForPatient(appointmentReportSearchDto.PatientId);
+
+            operations = SearchForDate(operations, appointmentReportSearchDto);
+
+            operations = SearchForDoctorNameAndSurname(operations, appointmentReportSearchDto);
+
+            operations = SearchForAppointmentType(operations, appointmentReportSearchDto);
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs:75]: Write in the form:
"List<Operation> operations = SearchForAppointmentType(SearchForDoctorNameAndSurname(SearchForDate(GetOperationsForPatient(appointmentReportSearchDto.PatientId), appointmentReportSearchDto), appointmentReportSearchDto), appointmentReportSearchDto);"

File: PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs [hunk 8]
```diff
@@ -55,7 +58,68 @@ public Boolean isTermNotAvailable(DoctorUser doctor, TimeSpan start, TimeSpan en
...
+
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs:76]: Delete the unnecessarily added blank line.

File: PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs [hunk 9]
```diff
@@ -55,7 +58,68 @@ public Boolean isTermNotAvailable(DoctorUser doctor, TimeSpan start, TimeSpan en
...
+            return operations;
+
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs:78]: Delete the unnecessarily added blank line.

File: PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs [hunk 10]
```diff
@@ -55,7 +58,68 @@ public Boolean isTermNotAvailable(DoctorUser doctor, TimeSpan start, TimeSpan en
...
+        }
+
+        private List<Operation> SearchForDoctorNameAndSurname(List<Operation> operations, AppointmentReportSearchDto appointmentSearchDto)
+        {
+            if (!appointmentSearchDto.DoctorNameAndSurname.Equals(""))
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs:83]: It is necessary to add a method that checks if it is an empty string, and replace it in the if statement.

File: PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs [hunk 11]
```diff
@@ -55,7 +58,68 @@ public Boolean isTermNotAvailable(DoctorUser doctor, TimeSpan start, TimeSpan en
...
+            {
+                operations = operations.FindAll(operation => operation.Doctor.firstName.Contains(appointmentSearchDto.DoctorNameAndSurname) || operation.Doctor.secondName.Contains(appointmentSearchDto.DoctorNameAndSurname));
+            }
+
+            return operations;
+        }
+
+        private List<Operation> SearchForDate(List<Operation> operations, AppointmentReportSearchDto appointmentSearchDto)
+        {
+            if (!appointmentSearchDto.Start.Equals("") && !appointmentSearchDto.End.Equals(""))
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs:93]: Apply a method that checks to see if it is an empty string.

File: PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs [hunk 12]
```diff
@@ -55,7 +58,68 @@ public Boolean isTermNotAvailable(DoctorUser doctor, TimeSpan start, TimeSpan en
...
+            {
+                DateTime startDate = DateTime.ParseExact(appointmentSearchDto.Start, "dd/MM/yyyy", CultureInfo.InvariantCulture);
+                DateTime endDate = DateTime.ParseExact(appointmentSearchDto.End, "dd/MM/yyyy", CultureInfo.InvariantCulture);
+                operations = operations.FindAll(operation => startDate <= DateTime.ParseExact(operation.Date, "dd/MM/yyyy", CultureInfo.InvariantCulture) && DateTime.ParseExact(operation.Date, "dd/MM/yyyy", CultureInfo.InvariantCulture) <= endDate);
+            }
+            else if (appointmentSearchDto.Start.Equals("") && !appointmentSearchDto.End.Equals(""))
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs:99]: Apply a method that checks to see if it is an empty string.

File: PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs [hunk 13]
```diff
@@ -55,7 +58,68 @@ public Boolean isTermNotAvailable(DoctorUser doctor, TimeSpan start, TimeSpan en
...
+            {
+                DateTime endDate = DateTime.ParseExact(appointmentSearchDto.End, "dd/MM/yyyy", CultureInfo.InvariantCulture);
+                operations = operations.FindAll(operation => DateTime.ParseExact(operation.Date, "dd/MM/yyyy", CultureInfo.InvariantCulture) <= endDate);
+            }
+            else if (!appointmentSearchDto.Start.Equals("") && appointmentSearchDto.End.Equals(""))
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs:104]: Apply a method that checks to see if it is an empty string.

File: PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs [hunk 14]
```diff
@@ -55,7 +58,68 @@ public Boolean isTermNotAvailable(DoctorUser doctor, TimeSpan start, TimeSpan en
...
+                operations = operations.FindAll(operation => startDate <= DateTime.ParseExact(operation.Date, "dd/MM/yyyy", CultureInfo.InvariantCulture));
+            }
+
+            return operations;
+
+        }
+
+        private List<Operation> SearchForAppointmentType(List<Operation> operations, AppointmentReportSearchDto appointmentSearchDto)
+        {
+            if (appointmentSearchDto.AppointmentType.Equals("") || appointmentSearchDto.AppointmentType.Equals("Operation"))
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/OperationService.cs:116]: Apply a method that checks to see if it is an empty string. Create a method that checks if the appointment type is an operation and apply it in the if statement.

File: PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs [hunk 15]
```diff
@@ -145,5 +153,62 @@ public Boolean isTermNotAvailable(DoctorUser doctor, TimeSpan time, String dateT
...
+        {
+            List<DoctorAppointment> appointments = GetAppointmentsForPatient(appointmentReportSearchDto.PatientId);
+
+            appointments = SearchForDate(appointments, appointmentReportSearchDto);
+
+            appointments = SearchForDoctorNameAndSurname(appointments, appointmentReportSearchDto);
+
+            appointments = SearchForAppointmentType(appointments, appointmentReportSearchDto);
+
+            return appointments;
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs]: Write in the form:
"return SearchForAppointmentType(SearchForDoctorNameAndSurname(SearchForDate(GetAppointmentsForPatient(appointmentReportSearchDto.PatientId), appointmentReportSearchDto), appointmentReportSearchDto), appointmentReportSearchDto);"

File: PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs [hunk 16]
```diff
@@ -145,5 +153,62 @@ public Boolean isTermNotAvailable(DoctorUser doctor, TimeSpan time, String dateT
...
+
+        }
+
+        private List<DoctorAppointment> SearchForDoctorNameAndSurname(List<DoctorAppointment> appointments, AppointmentReportSearchDto appointmentSearchDto)
+        {
+            if (!appointmentSearchDto.DoctorNameAndSurname.Equals(""))
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs:163]: It is necessary to add a method that checks if it is an empty string, and replace it in the if statement.

File: PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs [hunk 17]
```diff
@@ -145,5 +153,62 @@ public Boolean isTermNotAvailable(DoctorUser doctor, TimeSpan time, String dateT
...
+            {
+                appointments = appointments.FindAll(appointment => appointment.Doctor.firstName.Contains(appointmentSearchDto.DoctorNameAndSurname) || appointment.Doctor.secondName.Contains(appointmentSearchDto.DoctorNameAndSurname));
+            }
+
+            return appointments;
+        }
+
+        private List<DoctorAppointment> SearchForDate(List<DoctorAppointment> appointments, AppointmentReportSearchDto appointmentSearchDto)
+        {
+            if (!appointmentSearchDto.Start.Equals("") && !appointmentSearchDto.End.Equals(""))
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs:173]: Apply a method that checks to see if it is an empty string.

File: PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs [hunk 18]
```diff
@@ -145,5 +153,62 @@ public Boolean isTermNotAvailable(DoctorUser doctor, TimeSpan time, String dateT
...
+            {
+                DateTime startDate = DateTime.ParseExact(appointmentSearchDto.Start, "dd/MM/yyyy", CultureInfo.InvariantCulture);
+                DateTime endDate = DateTime.ParseExact(appointmentSearchDto.End, "dd/MM/yyyy", CultureInfo.InvariantCulture);
+                appointments = appointments.FindAll(appointment => startDate <= DateTime.ParseExact(appointment.Date, "dd/MM/yyyy", CultureInfo.InvariantCulture) && DateTime.ParseExact(appointment.Date, "dd/MM/yyyy", CultureInfo.InvariantCulture) <= endDate);
+            }
+            else if (appointmentSearchDto.Start.Equals("") && !appointmentSearchDto.End.Equals(""))
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs:179]: Apply a method that checks to see if it is an empty string.

File: PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs [hunk 19]
```diff
@@ -145,5 +153,62 @@ public Boolean isTermNotAvailable(DoctorUser doctor, TimeSpan time, String dateT
...
+            {
+                DateTime endDate = DateTime.ParseExact(appointmentSearchDto.End, "dd/MM/yyyy", CultureInfo.InvariantCulture);
+                appointments = appointments.FindAll(appointment => DateTime.ParseExact(appointment.Date, "dd/MM/yyyy", CultureInfo.InvariantCulture) <= endDate);
+            }
+            else if(!appointmentSearchDto.Start.Equals("") && appointmentSearchDto.End.Equals(""))
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs:184]: Apply a method that checks to see if it is an empty string.

File: PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs [hunk 20]
```diff
@@ -145,5 +153,62 @@ public Boolean isTermNotAvailable(DoctorUser doctor, TimeSpan time, String dateT
...
+                appointments = appointments.FindAll(appointment => startDate <= DateTime.ParseExact(appointment.Date, "dd/MM/yyyy", CultureInfo.InvariantCulture));
+            }
+            
+            return appointments;
+
+        }
+
+        private List<DoctorAppointment> SearchForAppointmentType(List<DoctorAppointment> appointments, AppointmentReportSearchDto appointmentSearchDto)
+        {
+            if (appointmentSearchDto.AppointmentType.Equals("") || appointmentSearchDto.AppointmentType.Equals("Appointment"))
```
Comment (trki293) [inline PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs:196]: Apply a method that checks to see if it is an empty string. Create a method that checks if the appointment type is an appointment and apply it in the if statement.

**Review by trki293 — CHANGES_REQUESTED** [review]
Corrections required in the code.

**Review by trki293 — APPROVED** [review]
PR is approved, good Clean code practice followed.

File: PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs [hunk 21]
```diff
@@ -145,5 +155,55 @@ public Boolean isTermNotAvailable(DoctorUser doctor, TimeSpan time, String dateT
...
+            {
+                DateTime endDate = DateTime.ParseExact(appointmentSearchDto.End, "dd/MM/yyyy", CultureInfo.InvariantCulture);
+                appointments = appointments.FindAll(appointment => DateTime.ParseExact(appointment.Date, "dd/MM/yyyy", CultureInfo.InvariantCulture) <= endDate);
+            }
+            else if(!UtilityMethods.CheckIfStringIsEmpty(appointmentSearchDto.Start) && UtilityMethods.CheckIfStringIsEmpty(appointmentSearchDto.End))
+            {
+                DateTime startDate = DateTime.ParseExact(appointmentSearchDto.Start, "dd/MM/yyyy", CultureInfo.InvariantCulture);
+                appointments = appointments.FindAll(appointment => startDate <= DateTime.ParseExact(appointment.Date, "dd/MM/yyyy", CultureInfo.InvariantCulture));
+            }
+            return appointments;
```
Comment (Nemanja0505) [inline PatientWebApplication/HealthClinic.DAL/Service/RegularAppointmentService.cs:190]: This method is preferable and it is necessary to refactor. Make several smaller methods where each will do one thing.

File: PatientWebApplication/PatientWebApplication/Validators/AppointmentReportSearchValidator.cs [hunk 22]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        public AppointmentReportSearchValidator()
+        {
+            var date = new DateTime();
+            When(f => !f.Start.Equals(""), () =>
+            {
+                RuleFor(f => DateTime.TryParseExact(f.Start, "dd/MM/yyyy", CultureInfo.InvariantCulture, DateTimeStyles.None, out date)).Equal(true);
+            });
+            When(f => !f.End.Equals(""), () => {
+                RuleFor(f => DateTime.TryParseExact(f.End, "dd/MM/yyyy", CultureInfo.InvariantCulture, DateTimeStyles.None, out date)).Equal(true);
```
Comment (Nemanja0505) [inline PatientWebApplication/PatientWebApplication/Validators/AppointmentReportSearchValidator.cs:21]: It is necessary to refactor. Put date parsing and verification of date is equals empty string into two separate methods.

**Review by Nemanja0505 — CHANGES_REQUESTED** [review]
PR cannot be approved if the clean code is not followed. The indicated parts of the code should be refactored.

**Review by Nemanja0505 — APPROVED** [review]
PR can be approved a clean code is followed.

**Pull request merged**